optional-goi-analysis/goi-calculations.R

```diff
@@ -0,0 +1,291 @@
+       You may want to run the core workflow analysis before re-running this GOI analysis.")
+}
+
+# Check for dimensionality reduction results
+if(!(any(c("PCA", "UMAP") %in% reducedDimNames(sce)))){
+  stop("There are no PCA or UMAP results in the provided SingleCellExperiment object.
+       You may want to run the core workflow analysis before re-running this GOI analysis.")
+}
+#### Perform mapping -----------------------------------------------------------
+
+if (opt$perform_mapping == TRUE) {
+  # turn column name into symbol for pulling the column info out of data frame
+  ids_for_mapping <- goi_list %>%
+    dplyr::pull(gene_id)
+  
+  # Define the annotation packages based on the specified organism
+  annotation_list <- list(
+    'Homo sapiens' = "org.Hs.eg.db",
+    'Mus musculus' = "org.Mm.eg.db",
+    'Danio rerio' = "org.Dr.eg.db"  #,
+    # new organisms would go here
+  )
+  # Error handling
+  if (!(opt$organism %in% names(annotation_list))) {
+    stop(paste(opt$organism, "is not supported"))
+  }
+  # Load library and assign annotation package
+  library(annotation_list[[opt$organism]], character.only = TRUE)
+  annotation_package <-
+    eval(parse(text = annotation_list[[opt$organism]]))
+  
+  # Check that provided keytypes are present in the annotation package
+  if (!(opt$provided_identifier %in% keytypes(annotation_package))) {
+    stop(paste(opt$provided_identifier, "is not a supported type of gene identifiers."))
+  }
+  
+  # Perform mapping
+  goi_list <- mapIds(
+    # organism annotation package
+    annotation_package,
+    # the provided gene identifiers
+    keys = ids_for_mapping,
+    # the type of provided gene identifiers
+    keytype = opt$provided_identifier,
+    # the type of gene identifiers to map to
+    column = opt$sce_rownames_identifier,
+    multiVals = opt$multi_mappings
+  ) %>%
+    tibble::enframe(name = 'gene_id',
+                    value = tolower(opt$sce_rownames_identifier)) %>%
+    # enframe() makes a `list` column; we will simplify it with unnest()
+    # This will result in one row of our data frame per list item
+    tidyr::unnest(cols = tolower(opt$sce_rownames_identifier)) %>%
+    # grab only the unique rows
+    dplyr::distinct() %>%
+    dplyr::left_join(goi_list, by = "gene_id")
+  
+  # Save mapped object to file
+  write_tsv(goi_list, file.path(
+    opt$output_directory,
+    paste0(opt$library_id, "_mapped_genes.tsv")
+  ))
+  
+  # define mapped goi associated with rownames of SCE object
+  goi_rownames_column <- tolower(opt$sce_rownames_identifier)
+  goi_rownames <- goi_list %>%
+    dplyr::pull(goi_rownames_column)
+  
+} else {
+  # if no mapping is performed then set goi to input id column
+  goi_rownames <- goi_list %>%
+    dplyr::pull(gene_id)
+}
+
+#### Prepare data for plotting -------------------------------------------------
+
+# get logcounts from normalized sce object
+normalized_sce_logcounts_matrix <- as.matrix(t(logcounts(sce)))
```

Comment:
```suggestion
normalized_logcounts_matrix <- as.matrix(t(logcounts(sce)))
```
I think throughout you could remove the `_sce_` from the variable names since these are now matrices and not SCE objects.